File: runme/mcgen/local/datan/exasrc/e2min.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <string.h>
#include "datsrc.h"
#include "grsrc.h"

/* Common Block Declarations */

struct cmings_ {
    double y[1000];
    integer nny;
} LIBDATA cmings_;

/* Table of constant values */

static integer c__1 = 1, c__2 = 2, c__3 = 3, c__4 = 4, c__74 = 74;
static double c_dmp2 = -.2, c_dp9 = .9, c_dp15 = .15, c_dp85 = .85,
              c_dmp414 = -.414, c_d1 = 1., c_d0 = 0.;

main(void)
{
    /* System generated locals */
    integer i__1;
    double d__1;

    /* Local variables */
    static integer i, k, ipl, npl, list[2], nstep, nseed1, nseed2, nws;
    static double fact, fmin, xmin, xmax, ymin, ymax, x[2],
                  scrat[6] /* was [2][3] */, scrat1[4] /* was [2][2] */,
                  cx[4] /* was [2][2] */, epsiln, dpl, xpl[100], ypl[100];
    static char string[75];
    static logical ok;

/* identify program to user */
    printf("%s\n", "Program E2MIN demonstrates use of MINCOV");
    printf("%s\n", "by fitting Gaussian to small sample");
    printf("%s\n\n", "and determining errors of parameters by MINCOV.");
/* ask for input */
    printf("%s\n", "Enter number n of events (2 <= n <= 1000):");
    printf("%s", "> ");
    scanf("%li", &cmings_.nny);
/* draw sample */
    nseed1 = 1;
    nseed2 = 2;
    rne2in_(&nseed1, &nseed2);
    rnstnr_(cmings_.y, &cmings_.nny);
/* write table of data */
    printf("%s\n", "sample Y is");
    mtxwrt_(cmings_.y, &c__1, &cmings_.nny);
/* set first approximation */
    x[0] = 1.;
    x[1] = 2.;
/* minimize with MINSIM */
    list[0] = 1;
    list[1] = 1;
    nstep = 10000;
    epsiln = 0.;
    fmin = .01;
    printf("%s\n", "minimization with MINSIM");
    printf("%s%2li%s%2li%s%2li%s%2li%2li\n", "N = ", cmings_.nny,
            ", NR = ", c__2, ", NRED = ", c__2, ", LIST = ", list[0], list[1]);
    printf("%s%10.5lf%10.5lf\n", "first approx.: X = ", x[0], x[1]);
    minsim_(x, &c__2, &c__2, list, mingls_, &fmin, &epsiln, &nstep, scrat);
    printf("%s%6.2lf%s%6li\n%s%10.5lf%10.5lf\n",
          "result of minimization: FMIN = ", fmin, ", NSTEP =", nstep,
          "X =", x[0], x[1]);
/* determine covariance matrix */
    fact = 1.;
    mincov_(x, &c__2, &c__2, list, mingls_, &fact, scrat, scrat1, cx, &ok);
    if (! ok) {
        printf("%s\n", "determination of covariance matrix fails");
        exit(0);
    }
    printf("%s\n", "covariance matrix CX = ");
    mtxwrt_(cx, &c__2, &c__2);
/* ask for number of workstation */
    grnbws_();
    printf("\n%s\n", "Please, enter number of workstation:");
    printf("%s", "> ");
    scanf("%li", &nws);
    (void)getchar();
/* plot sample as one dimensional scatter plot and Gaussian */
    xmin = -5.;
    xmax = 5.;
    ymin = 0.;
    ymax = .5;
    gropen_();
    gropws_(&nws);
    grwncc_(&xmin, &xmax, &ymin, &ymax);
    grvwwc_(&c_dmp2, &c_dp9, &c_dp15, &c_dp85);
    grwnwc_(&c_dmp414, &c_d1, &c_d0, &c_d1);
    grfram_();
    grboun_();
    sprintf(string, "N = %4li, x_1# = %6.3lf, x_2# = %6.3lf"
            ", &D@x_1# = %6.3lf, &D@x_2# = %6.3lf", cmings_.nny, x[0], x[1],
            sqrt(cx[0]), sqrt(cx[3]));
    string[74] = ' ';
    grtxtc_(&c_d1, string, &c__74);
    memset(string, ' ', 75);
    strncpy(string, "x_1", 3);
    grsclx_(string, &c__3);
    strncpy(string, "x_2", 3);
    grscly_(string, &c__3);
/* plot scatter diagram */
    grstcl_(&c__2);
    i__1 = cmings_.nny;
    for (i = 1; i <= i__1; ++i) {
        xpl[0] = cmings_.y[i - 1];
        xpl[1] = xpl[0];
        ypl[0] = 0.;
        ypl[1] = .1;
        grplin_(&c__2, xpl, ypl);
    }
/* draw Gaussian corresponding to solution */
    fact = .3989423 / x[1];
    npl = 100;
    dpl = (xmax - xmin) / (double) (npl - 1);
    i__1 = npl;
    for (ipl = 1; ipl <= i__1; ++ipl) {
        xpl[ipl - 1] = xmin + (double) (ipl - 1) * dpl;
/* Computing 2nd power */
        d__1 = (xpl[ipl - 1] - x[0]) / x[1];
        ypl[ipl - 1] = fact * exp(d__1 * d__1 * -.5);
    }
    grstcl_(&c__4);
    grplin_(&npl, xpl, ypl);
    grclse_();
    return 0;
} /* main */

File: runme/mcgen/local/datan/exasrc/e2min.f
      PROGRAM E2MIN
      IMPLICIT DOUBLE PRECISION(A-H,O-Z)
      COMMON /CMINGS/ Y,NNY
      DIMENSION X(2),Y(1000),XPL(100),YPL(100),LIST(2)
      DIMENSION CX(2,2),SCRAT(2,3),SCRAT1(2,2)
      PARAMETER(NR=2,NRED=2)
      DOUBLE PRECISION MINGLS
      CHARACTER*75 STRING
      LOGICAL OK
      EXTERNAL MINGLS
C identify program to user
      WRITE(*,*)' Program E2MIN demonstrates use of MINCOV'
      WRITE(*,*)' by fitting Gaussian to small sample'
      WRITE(*,*)' and determining errors of parameters by MINCOV'
      WRITE(*,*)' '
C ask for input
      WRITE(*,*)' Enter number n of events (2 <= n <=1000)'
      WRITE(*,*)' >'
      READ(*,*)NNY
C draw sample
      NSEED1=1
      NSEED2=2
      CALL RNE2IN(NSEED1,NSEED2)
      CALL RNSTNR(Y,NNY)
C write table of data
      WRITE(*,*)' sample Y is'
      CALL MTXWRT(Y,1,NNY) 
C set first approximation
      X(1)=1.D0
      X(2)=2.D0
C minimize with MINSIM
      LIST(1)=1
      LIST(2)=1
      NSTEP=10000
      EPSILN=0.D0
      FMIN=.01D0
      WRITE(*,*)' minimization with MINSIM'
      WRITE(*,'(3(A,I2),A,2I2)')' N = ',NNY,', NR = ',NR,
     +', NRED = ',NRED,', LIST = ',LIST
      WRITE(*,'(A,2F10.5)')' first approx.: X = ',X
      CALL MINSIM(X,NR,NRED,LIST,MINGLS,FMIN,EPSILN,NSTEP,SCRAT)
      WRITE(*,'(A,F6.2,A,I6,/,A,(2F10.5))')
     +' result of minimization: FMIN = ', FMIN,', NSTEP =',NSTEP,
     +' X =',X
C determine covariance matrix
      FACT=DBLE(1.)
      CALL MINCOV(X,NR,NRED,LIST,MINGLS,FACT,SCRAT,SCRAT1,CX,OK)
      IF(.NOT.OK) THEN
        WRITE(*,*)' determination of covariance matrix fails'
        STOP
      END IF
      WRITE(*,*)' covariance matrix CX = '
      CALL MTXWRT(CX,NRED,NRED)
      WRITE(*,'(/)')
C ask for number of workstation
      CALL GRNBWS()
      WRITE(*,*)' Please, enter number of workstation:'
      WRITE(*,*)' >'
      READ(*,*)NWS
C plot sample as one dimensional scatter plot and Gaussian
      XMIN=-5.D0
      XMAX=5.D0
      YMIN=0.D0
      YMAX=.5D0
      CALL GROPEN
      CALL GROPWS(NWS)
      CALL GRWNCC(XMIN,XMAX,YMIN,YMAX)
      CALL GRVWWC(-0.2D0,0.9D0,0.15D0,0.85D0)
      CALL GRWNWC(-.414D0,1.D0,0.D0,1.D0)
      CALL GRFRAM
      CALL GRBOUN
      STRING(1:38)='N = ****, x_1# = ******, x_2# = ******'
      STRING(39:74)=', &D@x_1# = ******, &D@x_2# = ******'
      WRITE(STRING(5:8),'(I4)') NNY
      WRITE(STRING(18:23),'(F6.3)') X(1)
      WRITE(STRING(33:38),'(F6.3)') X(2)
      WRITE(STRING(51:56),'(F6.3)') SQRT(CX(1,1))
      WRITE(STRING(69:74),'(F6.3)') SQRT(CX(2,2))
      CALL GRTXTC(1.D0,STRING,74)
      STRING='x_1'
      CALL GRSCLX(STRING,3)
      STRING='x_2'
      CALL GRSCLY(STRING,3)
C plot scatter diagram
      CALL GRSTCL(2)
      DO 20 I=1,NNY
        XPL(1)=Y(I)
        XPL(2)=XPL(1)
        YPL(1)=0.D0
        YPL(2)=0.1D0
        CALL GRPLIN(2,XPL,YPL)
20    CONTINUE
C draw Gaussian corresponding to solution
      FACT=.3989423D0/X(2)
      NPL=100
      DPL=(XMAX-XMIN)/DBLE(NPL-1)
      DO 30 IPL=1,NPL
        XPL(IPL)=XMIN+DBLE(IPL-1)*DPL
        YPL(IPL)=FACT*EXP(-0.5D0*((XPL(IPL)-X(1))/X(2))**2)
30    CONTINUE
      CALL GRSTCL(4)
      CALL GRPLIN(NPL,XPL,YPL)
      CALL GRCLSE
      END
